### PR TITLE
[rush] rush-sdk try to load from global folder before install-and-run rush

### PR DESCRIPTION
**Pull request by chengcyber**
## Summary

This PR add a feature to `rush-sdk`. 

## Details

Before invoking install-and-run rush, try to load rush-lib from rush global folder for a better reuse of installed rush-lib.

## How it was tested

Tested locally

## Impacted documentation

Nope

### CODE REVIEW

File: libraries/rush-sdk/src/RushGlobalFolder.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    } else {
+      this.path = path.join(getHomeFolder(), '.rush');
+    }
+
+    const normalizedNodeVersion: string = process.version.match(/^[a-z0-9\-\.]+$/i)
+      ? process.version
+      : 'unknown-version';
+    this.nodeSpecificPath = path.join(this.path, `node-${normalizedNodeVersion}`);
+  }
+}
```
Comment (iclanton) [inline libraries/rush-sdk/src/RushGlobalFolder.ts:36]: Instead of copying this from `rush-lib`, why don't we deep-import this from rush-lib's `lib-esm` folder and bundle all of `rush-sdk`?

File: libraries/rush-sdk/src/RushGlobalFolder.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * @privateRemarks
+ * Keep this in sync with `EnvironmentConfiguration._getRushGlobalOverride` in rush.
+ */
+function getRushGlobalFolderOverride(processEnv: Record<string, string | undefined>): string | undefined {
+  const value: string | undefined = processEnv.RUSH_GLOBAL_FOLDER;
+  if (value) {
+    const normalizedValue: string | undefined = normalizeDeepestParentFolderPath(value);
+    return normalizedValue;
+  }
+}
```
Comment (iclanton) [inline libraries/rush-sdk/src/RushGlobalFolder.ts:52]: Keeping with the bundling idea, it'd be useful to stick this in its own module in `rush-lib` so it can get easily bundled.

File: libraries/rush-sdk/src/index.ts [hunk 3]
```diff
@@ -105,50 +111,67 @@ if (rushLibModule === undefined) {
...
+        terminal.writeVerboseLine(`Trying to load  ${RUSH_LIB_NAME} installed by install-run-rush`);
+        rushLibModule = requireRushLibUnderFolderPath(installRunNodeModuleFolder);
+      } catch (e) {
+        let installAndRunRushStderrContent: string = '';
+        try {
+          // Third, try to invoke install-run-rush.js to install the appropriate version of "rush-lib"
+          const installAndRunRushJSPath: string = path.join(
+            monorepoRoot,
+            'common/scripts/install-run-rush.js'
+          );
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:143]:
```suggestion
          const installAndRunRushJSPath: string = `${monorepoRoot}/common/scripts/install-run-rush.js`;
```

File: common/changes/@microsoft/rush/feat-rush-sdk-load-from-global-folder_2023-01-08-12-28.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "rush-sdk will try to load from rush global folder before install-run-rush",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/feat-rush-sdk-load-from-global-folder_2023-01-08-12-28.json:5]:
```suggestion
      "comment": "Add functionality to `@rushstack/rush-sdk` to try to load Rush from the global installation location before installing it by invoking the `install-run-rush` script.",
```

File: libraries/rush-sdk/src/RushGlobalFolder.ts (hunk 1 above)
Comment (chengcyber) [reply libraries/rush-sdk/src/RushGlobalFolder.ts:36]: Hi @iclanton , could you elaborate a little bit about this idea? 

Are you talking about

1. make `rush-sdk` dependency over `rush-lib`
2. introduce webpack.config.js to bundle `rush-sdk`
3. use webpack deep-import plugin to handle `rush-lib` imports

I know copying and syncing is bad. I just want to better understand your proposal.

Comment (iclanton) [reply libraries/rush-sdk/src/RushGlobalFolder.ts:36]: Yeah that's basically what I had in mind. `rush-sdk` already has a devDependency on `rush-lib`.

File: libraries/rush-sdk/src/index.ts [hunk 5]
```diff
@@ -13,6 +13,8 @@ import {
   Terminal,
   ConsoleTerminalProvider
 } from '@rushstack/node-core-library';
+import { RushGlobalFolder } from '@microsoft/rush-lib/lib/api/RushGlobalFolder';
```
Comment (dmichon-msft) [inline libraries/rush-sdk/src/index.ts:16]: Make sure this imports from whichever folder has the ESM output.

File: libraries/rush-sdk/webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      new DeepImportsPlugin({
+        path: `${__dirname}/temp/rush-sdk-manifest.json`,
+        inFolderName: 'lib-esnext',
+        outFolderName: 'lib',
+        pathsToIgnore: [],
+        dTsFilesInputFolderName: 'lib-commonjs'
+      })
+    ],
+    externals: [
+      ({ request }, callback) => {
```
Comment (dmichon-msft) [inline libraries/rush-sdk/webpack.config.js:49]: This rule needs to expressly exclude `@microsoft/rush-lib` or else it will be trying to load `rush-lib` in order to figure out how to load `rush-lib`.

**Comment by chengcyber** [comment]
This MR has been a while. I recalled it and updated the code!

Use `webpack5` to bundle `@rushstack/rush-sdk` package. `RushGlobalFolder` from `rush-lib` will be bundled into the output code.

File: libraries/rush-sdk/config/heft.json [hunk 7]
```diff
@@ -32,13 +32,20 @@
...
-        "generate-stubs": {
+        "webpack": {
           "taskDependencies": ["typescript"],
+          "taskPlugin": {
+            "pluginPackage": "@rushstack/heft-webpack5-plugin"
+          }
+        },
+
+        "generate-stubs": {
+          "taskDependencies": ["webpack"],
```
Comment (iclanton) [inline libraries/rush-sdk/config/heft.json:43]:
```suggestion
          "taskDependencies": ["typescript"],
```

File: libraries/rush-sdk/tsconfig.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "extends": "./node_modules/local-node-rig/profiles/default/tsconfig-base.json",
   "compilerOptions": {
-    "outDir": "lib-shim",
+    "outDir": "lib-commonjs",
```
Comment (iclanton) [inline libraries/rush-sdk/tsconfig.json:4]: This should go in a folder called something like `lib-shim-esnext` and provide ES-style modules to Webpack.

File: libraries/rush-sdk/webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      chunkFilename: 'chunks/[name].js',
+      library: {
+        type: 'commonjs2'
+      }
+    },
+    target: 'node',
+    plugins: [
+      new PreserveDynamicRequireWebpackPlugin(),
+      new webpack.ids.DeterministicModuleIdsPlugin({
+        maxLength: 6
```
Comment (iclanton) [inline libraries/rush-sdk/webpack.config.js:36]: This option is probably unnecessary.

File: libraries/rush-sdk/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
+/* eslint-env es6 */
```
Comment (iclanton) [inline libraries/rush-sdk/webpack.config.js:1]: This can happen in a follow-up, but a lot of this should be able to be deduplicated with the webpack config in rush-lib.

File: libraries/rush-sdk/src/index.ts [hunk 11]
```diff
@@ -126,51 +131,78 @@ if (sdkContext.rushLibModule === undefined) {
...
-
     try {
-      // First, try to load the version of "rush-lib" that was installed by install-run-rush.js
-      terminal.writeVerboseLine(`Trying to load  ${RUSH_LIB_NAME} installed by install-run-rush`);
-      sdkContext.rushLibModule = requireRushLibUnderFolderPath(installRunNodeModuleFolder);
-    } catch (e1) {
-      let installAndRunRushStderrContent: string = '';
-      try {
-        const installAndRunRushJSPath: string = path.join(monorepoRoot, 'common/scripts/install-run-rush.js');
+      const { RushGlobalFolder } = require('@microsoft/rush-lib/lib-esnext/api/RushGlobalFolder');
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:135]: Import this at the top of the file with a normal `import`.

File: libraries/rush-sdk/src/index.ts [hunk 12]
```diff
@@ -126,51 +131,78 @@ if (sdkContext.rushLibModule === undefined) {
...
+      terminal.writeLine(RushGlobalFolder);
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:136]: Remove?

File: libraries/rush-sdk/src/index.ts [hunk 13]
```diff
@@ -126,51 +131,78 @@ if (sdkContext.rushLibModule === undefined) {
...
+      terminal.writeVerboseLine(`Try to load ${RUSH_LIB_NAME} from rush global folder`);
+      const rushGlobalFolder: typeof RushGlobalFolder = new RushGlobalFolder();
+      // The path needs to keep align with the logic inside RushVersionSelector
+      const expectedGlobalRushInstalledFolder: string = path.join(
+        rushGlobalFolder.nodeSpecificPath,
+        `rush-${rushVersion}`
+      );
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:143]:
```suggestion
      const expectedGlobalRushInstalledFolder: string = `${rushGlobalFolder.nodeSpecificPath}/rush-${rushVersion}`
```

File: libraries/rush-sdk/src/index.ts [hunk 14]
```diff
@@ -126,51 +131,78 @@ if (sdkContext.rushLibModule === undefined) {
...
-      sdkContext.rushLibModule = requireRushLibUnderFolderPath(installRunNodeModuleFolder);
-    } catch (e1) {
-      let installAndRunRushStderrContent: string = '';
-      try {
-        const installAndRunRushJSPath: string = path.join(monorepoRoot, 'common/scripts/install-run-rush.js');
+      const { RushGlobalFolder } = require('@microsoft/rush-lib/lib-esnext/api/RushGlobalFolder');
+      terminal.writeLine(RushGlobalFolder);
+      terminal.writeVerboseLine(`Try to load ${RUSH_LIB_NAME} from rush global folder`);
+      const rushGlobalFolder: typeof RushGlobalFolder = new RushGlobalFolder();
+      // The path needs to keep align with the logic inside RushVersionSelector
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:139]: Can that logic just be referenced here?

File: libraries/rush-sdk/src/index.ts [hunk 15]
```diff
@@ -126,51 +131,78 @@ if (sdkContext.rushLibModule === undefined) {
...
-        terminal.writeLine('The Rush engine has not been installed yet. Invoking install-run-rush.js...');
+    if (sdkContext.rushLibModule !== undefined) {
+      // to track which scenario is active and how it got initialized.
+      global.___rush___rushLibModuleFromRushGlobalFolder = sdkContext.rushLibModule;
+      terminal.writeVerboseLine(`Loaded ${RUSH_LIB_NAME} installed from rush global folder`);
+    } else {
+      const installRunNodeModuleFolder: string = path.join(
+        monorepoRoot,
+        `common/temp/install-run/@microsoft+rush@${rushVersion}`
+      );
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:160]:
```suggestion
      const installRunNodeModuleFolder: string = `${monorepoRoot}/common/temp/install-run/@microsoft+rush@${rushVersion}`
```

File: libraries/rush-sdk/src/index.ts [hunk 16]
```diff
@@ -21,7 +21,9 @@ import {
   sdkContext
 } from './helpers';
 
-const verboseEnabled: boolean = typeof process !== 'undefined' && process.env.RUSH_SDK_DEBUG === '1';
+const verboseEnabled: boolean =
+  typeof process !== 'undefined' &&
+  (process.env.RUSH_SDK_DEBUG === '1' || process.env._RUSH_SDK_DEBUG === '1');
```
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:26]: Why do we need a second env var?

Comment (chengcyber) [reply libraries/rush-sdk/src/index.ts:26]: `RUSH_SDK_DEBUG` in disallowed in rush.js, which means it's impossible to run `RUSH_SDK_DEBUG=1 rush command-invoke-rush-sdk`.

A new environment variable with a prefix `_` make it possible to run `_RUSH_SDK_DEBUG=1 rush my-invoke-rush-sdk` to show debug loggings.

File: libraries/rush-sdk/tsconfig.json (hunk 8 above)
Comment (chengcyber) [reply libraries/rush-sdk/tsconfig.json:4]: By following rush-lib's fashion, configuring `additionalModuleKindsToEmit: [ { moduleKind: 'esnext', { outFolderName: 'lib-esnext ' } ]`, and update rush-sdk's webpack.config.js to use `lib-esnext` folder as entries.

File: libraries/rush-sdk/webpack.config.js (hunk 9 above)
Comment (chengcyber) [reply libraries/rush-sdk/webpack.config.js:36]: Removed

File: libraries/rush-sdk/src/index.ts (hunk 14 above)
Comment (chengcyber) [reply libraries/rush-sdk/src/index.ts:139]: Possible.

But the logic is located in `rush` instead of `rush-lib`.  https://github.com/microsoft/rushstack/blob/8ed5cea52f35a93da97b54a8b7e0465be4b2f34c/apps/rush/src/RushVersionSelector.ts#L31

By directly referencing, `rush-sdk` needs to depends on `rush` as well.

File: libraries/rush-sdk/tsconfig.json (hunk 8 above)
Comment (chengcyber) [reply libraries/rush-sdk/tsconfig.json:4]: Unfornunately, using `lib-esnext` as entries don't work well with `rush-sdk`. In scenario 2, rush-sdk uses `module.parent` to read the parent module filename. This doesn't work in ESM code... which breaks the test cases for rush plugins in this repo. So, I reverted the webpack.config.js to use commonjs code as entries.

**Comment by chengcyber** [comment]
Hi @iclanton , I pushed code modifications as your suggestions. Could you kindly take another look? Thx!

File: libraries/rush-sdk/src/typings.d.ts [hunk 17]
```diff
@@ -0,0 +1 @@
+declare module '@microsoft/rush-lib/lib-esnext/api/RushGlobalFolder';
```
Comment (iclanton) [inline libraries/rush-sdk/src/typings.d.ts:1]: Why is this necessary?

**Review by iclanton — APPROVED** [review]

File: libraries/rush-sdk/src/typings.d.ts (hunk 17 above)
Comment (chengcyber) [reply libraries/rush-sdk/src/typings.d.ts:1]: If I removed this declaration, tsc for `rush-sdk` complains the following error. I suppose types are not generated under `rush-lib`'s `lib-esnext` folder.

CleanShot 2024-06-24 at 11 00 55@2x

Comment (dmichon-msft) [reply libraries/rush-sdk/src/typings.d.ts:1]: This can be solved by fiddling with `typesVersions` in package.json in rush-lib:
```json
"typesVersions": {
    "*": {
      "lib-esnext/*": [
        "lib/*"
      ]
    }
  }
```
Or similar mapping.

Comment (chengcyber) [reply libraries/rush-sdk/src/typings.d.ts]: Thanks @dmichon-msft "typesVersions" works like a charm!

I tested it locally and `tsc` can handle `rush-lib`'s deep imports well after adding `typesVerions` in package.json of `rush-lib`

Code updated to reflect this change.

cc: @iclanton

**Comment by chengcyber** [comment]
I merged the latest main branch yesterday. However, there is a merge conflicts with main's repo-state.json today... Do I need to merge the latest main branch again?

**Comment by octogonz** [comment]
🚀 Finally released with Rush 5.129.7
